### PR TITLE
New schedulers

### PR DESCRIPTION
**Pull request by jeremyasapp**
- WarmupLinearScheduler
- NoamScheduler

### CODE REVIEW

File: flambe/learn/distillation.py [hunk 1]
```diff
@@ -64,6 +65,8 @@ def __init__(self,
             The optimizer to use
         scheduler : torch.optim.lr_scheduler._LRScheduler, optional
             An optional learning rate scheduler
+        iter_scheduler : torch.optim.lr_scheduler._LRScheduler, optional
+            An optional learning rate scheduler to run after each ite
```
Comment (iitzco-asapp) [inline flambe/learn/distillation.py]: `iteration` typo

File: flambe/learn/train.py [hunk 2]
```diff
@@ -219,14 +219,9 @@ def _train_step(self) -> None:
 
                 # Update iter scheduler
                 if self.iter_scheduler is not None:
-                    log(f'{tb_prefix}Training/LR', self.iter_scheduler.get_lr()[0], global_step)
-                    if isinstance(self.iter_scheduler, ReduceLROnPlateau):
-                        self.iter_scheduler.step(val_loss)
```
Comment (iitzco-asapp) [inline flambe/learn/train.py]: Isn't this case valid anymore?

Comment (jeremyasapp) [reply flambe/learn/train.py]: On each iteration you can't look at the val loss really. We could expand to provide a closure based on the training loss I guess? Maybe let's keep it like this for now until the need arises

File: flambe/nn/mlp.py [hunk 3]
```diff
@@ -34,10 +34,11 @@ def __init__(self,
             Input_dimension
         output_size: int
             Output dimension
-        n_layers: int
+        n_layers: int, optional
             Number of layers in the network, defaults to 1
         dropout: float, optional
-            Dropout to be used before each MLP layer
+            Dropout to be used before each MLP layer.
+            Only used if n_layers > 1.
```
Comment (iitzco-asapp) [inline flambe/nn/mlp.py]: Nice detail.

File: flambe/optim/linear.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            The number of linear warmup phases
+        n_steps : int, optional
+            The index of last step. Default: -1
+
+        """
+        self.warmup = warmup
+        self.n_steps = n_steps
+        super().__init__(optimizer, lr_lambda=self.lr_lambda, last_epoch=n_steps)  # type: ignore
+
+    def lr_lambda(self, step: int) -> float:
```
Comment (iitzco-asapp) [inline flambe/optim/linear.py]: Lets add unit tests for this method.

Comment (iitzco-asapp) [reply flambe/optim/linear.py]: I see some risk in touching this code without being careful

File: flambe/optim/noam.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            The number of linear warmup phases
+        d_model : int, optional
+            The index of last step. Default: -1
+
+        """
+        self.warmup = warmup
+        self.d_model = d_model
+        super().__init__(optimizer, lambda_lr=self.lr_lambda, last_epoch=-1)  # type: ignore
+
+    def lr_lambda(self, step: int) -> float:
```
Comment (iitzco-asapp) [inline flambe/optim/noam.py]: Same here, lets add some tests

File: flambe/learn/train.py (hunk 2 above)
Comment (iitzco-asapp) [reply flambe/learn/train.py]: Does this mean that we will not support `ReduceLROnPlateau`?

Comment (jeremyasapp) [reply flambe/learn/train.py]: We do for the normal val step schedulers but it doesn't make sense for the iter scheduler
